neigh.go

```diff
@@ -0,0 +1,259 @@
+	b[1] = 0 //reserved
+	nlenc.PutUint16(b[2:4], m.Family)
+	nlenc.PutUint32(b[4:8], m.Index)
+	nlenc.PutUint16(b[8:10], m.State)
+	b[10] = m.Flags
+	b[11] = m.Type
+
+	if m.Attributes != nil {
+		a, err := m.Attributes.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+
+		return append(b, a...), nil
+	}
+	return b, nil
+}
+
+// UnmarshalBinary unmarshals the contents of a byte slice into a NeighMessage.
+func (m *NeighMessage) UnmarshalBinary(b []byte) error {
+	l := len(b)
+	if l < neighMsgLen {
+		return errInvalidNeighMessage
+	}
+
+	m.Family = nlenc.Uint16(b[2:4])
+	m.Index = nlenc.Uint32(b[4:8])
+	m.State = nlenc.Uint16(b[8:10])
+	m.Flags = b[10]
+	m.Type = b[11]
+
+	if l > neighMsgLen {
+		m.Attributes = &NeighAttributes{}
+		err := m.Attributes.UnmarshalBinary(b[neighMsgLen:])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// rtMessage is an empty method to sattisfy the Message interface.
+func (*NeighMessage) rtMessage() {}
+
+// NeighService is used to retrieve rtnetlink family information.
+type NeighService struct {
+	c *Conn
+}
+
+// New creates a new interface using the LinkMessage information.
+func (l *NeighService) New(req *NeighMessage) error {
+	flags := netlink.Request | netlink.Create | netlink.Acknowledge | netlink.Excl
+	_, err := l.c.Execute(req, unix.RTM_NEWNEIGH, flags)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Delete removes an neighbor entry by index.
+func (l *NeighService) Delete(index uint32) error {
+	req := &NeighMessage{}
+
+	flags := netlink.Request | netlink.Acknowledge
+	_, err := l.c.Execute(req, unix.RTM_DELNEIGH, flags)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// List retrieves all neighbors.
+func (l *NeighService) List() ([]NeighMessage, error) {
+	req := &NeighMessage{}
+
+	flags := netlink.Request | netlink.Dump
+	msgs, err := l.c.Execute(req, unix.RTM_GETNEIGH, flags)
+	if err != nil {
+		return nil, err
+	}
+
+	neighs := make([]NeighMessage, 0, len(msgs))
+	for _, m := range msgs {
+		neigh := (m).(*NeighMessage)
+		neighs = append(neighs, *neigh)
+	}
+
+	return neighs, nil
+}
+
+// NeighCacheInfo contains neigh information
+type NeighCacheInfo struct {
+	Confirmed uint32
+	Used      uint32
+	Updated   uint32
+	RefCount  uint32
+}
+
+// UnmarshalBinary unmarshals the contents of a byte slice into a NeighMessage.
+func (n *NeighCacheInfo) UnmarshalBinary(b []byte) error {
+	if len(b) != 16 {
+		return fmt.Errorf("incorrect size, want: 16, got: %d", len(b))
+	}
+
+	n.Confirmed = nlenc.Uint32(b[0:4])
+	n.Used = nlenc.Uint32(b[4:8])
+	n.Updated = nlenc.Uint32(b[8:12])
+	n.RefCount = nlenc.Uint32(b[12:16])
+
+	return nil
+}
+
+// NeighAttributes contains all attributes for a neighbor.
+type NeighAttributes struct {
+	Address   net.IP           // a neighbor cache n/w layer destination address
+	LLAddress net.HardwareAddr // a neighbor cache link layer address
+	CacheInfo *NeighCacheInfo  // cache statistics
+	IfIndex   uint32
+}
+
+const (
```

Comment:
Are these not available in the Unix package?

Reply:
Unfortunately no. I could not find any of them in `golang.org/x/sys/unix`